### ISSUE TITLE
Provide entrypoint for successful pipx installation

### ISSUE THREAD

**Issue by Gitznik** [ranger/ranger#2941]
#### Runtime Environment

- Operating system and version: macos 13.6.3
- Terminal emulator and version: macos terminal
- Python version: 3.12.1
- Ranger version/commit: 1.9.3
- Locale: 

#### Current Behavior

Hey there,

currently this tool cannot be invoked via pipx, if pipx bases its virtual environments in a path that contains spaces, which is the default on mac, as per the mac documentation. This is the case since pipx 1.3. 

The solution is to provide a entry point, see the [pipx docs](https://pipx.pypa.io/stable/how-pipx-works/#developing-for-pipx).

As you recommend installing this tool via pipx, it would be great if you could update your package accordingly.

Thanks!

based on https://github.com/pypa/pipx/issues/1212

#### Expected Behavior

Entrypoint available so the package can be installed correctly, even if the venv is in a path with spaces.

#### Context

This was reported in pipx under this issue: https://github.com/pypa/pipx/issues/1226

#### Possible Solutions
Provide an entrypoint definition.

#### Steps to reproduce
1. Set the `PIPX_HOME` environment variable to a path with spaces
2. `pipx install ranger-fm`
3. `ranger` -> fails because the shebang of the ranger executable contains a space

#### Traceback

```

```

**Comment by toonn** [comment]
This strictly sounds like a Pipx issue. It should escape paths it injects properly.

If this remains an issue I'd drop the Pipx recommendation instead.

**Comment by Gitznik** [comment]
It's a `pip` issue, actually. `pipx` has nothing to do with these files. This problem also occurs if a user installs `ranger` into a `venv` that resides in a path with spaces.

We've since moved the default pipx install location on mac to a path  without spaces and print a warning whenever a user has installed pipx in a path with spaces.

**Comment by flying-sheep** [comment]
OK, I figured it out. The issue is
- in ranger: you shouldn’t use setuptools’ `scripts` for python scripts you expect to be run with a venv’s Python. use entry points instead
- in setuptools: it shouldn’t touch the shebang
- in pip: it shouldn‘t touch the shebang

In short: `scripts` is a distutils feature that hasn’t been standardized, so distutils’ behavior should be used, which is [simple](https://github.com/youknowone/python-deadlib/blob/815dfe935c90ec7180fdccc9710bf14032bdf3cb/distutils/distutils/command/install_scripts.py#L44):

> ```py
> self.outfiles = self.copy_tree(self.build_dir, self.install_dir)
> ```

Long version:

Setuptools’ `scripts` argument (used by ranger) is documented as being for *standalone* scripts. Now I would interpret that as files that are supposed to be copied into the target’s $PATH unmodified. Indeed someone who was probably a ranger maintainer (when the original bug was filed on bitbucket) [asked for exactly that](https://github.com/pypa/setuptools/issues/444), also as said above, `distutils` defines its behavior and defines it as “just copy things over unmodified”.

What happens is in details:
1. when building a wheel, setuptools changes the shebang from `#!/usr/bin/python` to `#!python -O` and puts it into the `.data/scripts` directory:

	```bash
	$ pip download ranger-fm
	$ uv build --wheel ranger_fm-*.tar.gz
	$ bsdtar tf ranger_fm-*-py3-none-any.whl | grep scripts
	ranger_fm-1.9.4.data/scripts/ranger
	ranger_fm-1.9.4.data/scripts/rifle
	```

2. `pip`, when installing into a venv, modifies the shebang again to point to the venv Python

As said, the `scripts` kwarg and the corresponding `.data/scripts` directory are implementation defined. And the implementation (distutils) doesn’t adapt shebangs at all.

**Comment by flying-sheep** [comment]
OK, I figured it out. The issue is in pip: it should treat the shebang in the same way it treats other shebangs

In short: `scripts` is a distutils feature that hasn’t been standardized, so distutils’ behavior should be used, which is https://github.com/youknowone/python-deadlib/blob/815dfe935c90ec7180fdccc9710bf14032bdf3cb/distutils/distutils/command/build_scripts.py#L54-L58

> Copy each script [in the .data/scripts directory]; if it's marked as a
> Python script in the Unix way (first line matches 'first_line_re',
> i.e. starts with "\#!" and contains "python"), then adjust the first
> line to refer to the [target] Python interpreter as we copy.

What happens is in details:
1. when building a wheel, setuptools changes the shebang from `#!/usr/bin/python` to `#!python -O` and puts it into the `.data/scripts` directory:

	```bash
	$ pip download ranger-fm
	$ uv build --wheel ranger_fm-*.tar.gz
	$ bsdtar tf ranger_fm-*-py3-none-any.whl | grep scripts
	ranger_fm-1.9.4.data/scripts/ranger
	ranger_fm-1.9.4.data/scripts/rifle
	```

2. `pip`, when installing into a venv, modifies the shebang again to point to the venv’s Python.

The issue is in how it does that, because
- `_internal.operations.install.wheel:fix_script` *is not as sophisticated* as
- `_vendor.distlib.script:ScriptMaker.get_shebang`

so pip should reuse the latter code in the former function.

----

PS: funnily, someone who is probably the maintainer of this repo (unclear after setuptools’ migration from bitbucket), asked for setuptools to *not* touch the shebang earlier: https://github.com/pypa/setuptools/issues/444

**Comment by flying-sheep** [comment]
So:
- pip _could_ reuse the latter code in the former function: https://github.com/pypa/pip/issues/13389
- pipx _could_ use `uv` as installer if a package contains `*.data/scripts` until that’s fixed (or just wait until pip fixes that
- ranger **should** just start using regular entry-point scripts

**Comment by flying-sheep** [comment]
No, you should definitely move away from using `(data)/scripts` for Python scripts and use `[project.scripts]` (i.e. entry-point based scripts) instead.

While not explicitly deprecated, others in the PyPA community have made clear that `(data)/scripts` is underspecified and not the way to go for Python scripts.

**Comment by flying-sheep** [comment]
One more thing: rewriting the shebang for `(data)/scripts` is only a “[recommended installer feature](https://packaging.python.org/en/latest/specifications/binary-distribution-format/#recommended-installer-features)” to you can’t rely on that happening.

**Comment by flying-sheep** [comment]
At this point, ranger is pretty much the only project I’m aware of that still relies on this, so pipx might just move on to be able to use standard directories. You should really fix that.

**Comment by flying-sheep** [comment]
@toonn the next version of pipx will use standard directories again, which means spaces in the path on macOS.

As said, you're relying on a “recommend installer feature”, not something that's guaranteed to work and you should switch to regular entry points.

Unless you do that, people who use pipx for the first time starting from the next pipx release will have a problem running ranger.

**Comment by hut** [comment]
Can you think of any way to use an entry point script only when installed via pipx?

Or any other solution that would maintain the extra feature when installed with other package managers?

**Comment by hut** [comment]
Also, @flying-sheep, thank you for nagging us with this. It's got to be annoying. I appreciate it.

**Comment by flying-sheep** [comment]
> Can you think of any way to use an entry point script only when installed via pipx?
>
> Or any other solution that would maintain the extra feature when installed with other package managers?

I think I forgot what extra feature exactly you’re referring to. From reading the script, maybe you mean that people are able to `source` that script with bash and it’ll do one thing, while when executed normally (by typing `ranger` when it’s installed in the $PATH), it’ll do another thing. Is that it?

If so, I’m not sure that can be preserved 100%, as it relies on a few implementation details (e.g. that and how exactly the shebang is re-written, none of which is guaranteed). I think you could investigate other methods how to do it, e.g.

- you could have `ranger` check if it’s executed from `pipx`’s bin directory, and if so, replace itself with the script (leaving the shebang alone)
- you could take up my pip PR and finish it so pip rewrites script shebangs in a way that works for you
- you could do something like `pipx` itself does, and add a `ranger add-alias <bash|zsh|...>` command to the CLI that can be used to append shell code to the user’s .bashrc/.zshrc/... and tells them to restart the shell. Users could then use `ranger cd` to change directories, with the shell code added being sth. like this:

  ```python
  script = f"""
  ranger() {
    if [[ "$1" -eq 'cd' ]]; then
      cd "$("{sys.executable}" -m ranger --print-end-path "$@")"  # or however that works
    else
      "{sys.executable}" -m ranger "$@"
    fi
  }
  """
  ```

  Linux distributions that package ranger would then add this code to the global `/etc/bashrc.d` or so directory.

To answer your question directly: all package managers are designed to follow standards in the same way, so I doubt you can get different behavior in them.

> Also, @flying-sheep, thank you for nagging us with this. It's got to be annoying. I appreciate it.

Thanks, I know on your end, changes like that can be a pain in the ass! But what motivates me is that back in the day, nobody thought Python packaging was straightforward or easy, while for a few years now, I haven’t heard this sentiment from newbies anymore. This is only possible because of standards work, and standardization means changes to things that exist for longer.

**Comment by hut** [comment]
Yes, exactly, that's the feature that I mean. `. ranger` would let you change directories on quit.

Maybe it's time to retire the feature. It *is* rather 1337, but probably not worth battling against standardization over.

@toonn @vifon what do you think?

**Comment by flying-sheep** [comment]
I just noticed something – pip’s robust shebang rewriting uses the same dual bash/python script trick, so applying that rewrite to your script would probably break it anyway.

So this only works
- for bash (maybe zsh, certainly not nushell, csh, …)
- if pipx happens to install scripts to a path with no spaces
- if pip uses its more fragile shebang rewriting strategy (I think)

Yeah, definitely built on sand.

You *could* make it more robust by implementing that `ranger add-alias` command as I suggested.

**Comment by toonn** [comment]
While `. ranger` is not robust to this shebang rewriting, it's a very popular feature. So I'm in favor of preserving it even if it won't work with Pipx.

**Comment by flying-sheep** [comment]
Yes, there’s one good place for that: Linux package managers. If it doesn’t work when installed with them, you should file bugs, and ideally provide install instructions how to install it correctly so the shebang works or ship a script that belongs into `/etc/profile.d` that sets things up. Then people can install it using `apt install ranger` or `pacman -S ranger` or so and get the feature.

E.g. it currently works on Arch Linux, `pacman -S` ranger leaves the shebang intact.

Everything else is a hack:

- pip: is not a tool manager, shouldn’t be used for global/userwide tools, Linux distributions started discouraging using a userwide env for stuff like that
- pipx and uv: if it works it’s incidental since there’s no guarantee that the shebang is written in a compatible way

**Comment by hut** [comment]
Sounds good.

@flying-sheep would you be open to write a PR for this? You seem to understand the situation the best.

**Comment by flying-sheep** [comment]
I think the biggest payoff is you going through the most important installers and checking if it works:

- homebrew
- nix
- debian
- fedora

And if not, filing an issue report.

If you want it to work in other shells, you’d have to design something, as `.` doesn’t make much sense in shells where `.` isn’t a synonym for `source`.
